backend/src/openarchiefbeheer/accounts/api/serializers.py

```diff
@@ -34,15 +35,49 @@ class Meta:
 
     @extend_schema_field(RoleSerializer)
     def get_role(self, user: User) -> dict | None:
-        data = {}
-        for permission in Permission.objects.filter(user=user):
-            data[permission.codename] = True
 
-        for group in user.groups.all():
-            for permission in group.permissions.all():
-                data[permission.codename] = True
+        # Retrieve all permission codenames.
+        permissions = []
 
-        serializer = RoleSerializer(data=data)
-        serializer.is_valid()
+        try:
+            """
+            Annotating a (User) queryset with:
 
-        return serializer.data
+              - `user_permission_codenames`
+              - `group_permission_codenames`
+
+            can improve performance by reducing the number of queries needed.
+
+            Example:
+
+            queryset.annotate(
+                user_permission_codenames=ArrayAgg(
+                    "user_permissions__codename",
+                    distinct=True,
+                ),
+                group_permission_codenames=ArrayAgg(
+                    "groups__permissions__codename",
+                    distinct=True,
+                ),
+            )
+            """
+            permissions += user.user_permission_codenames
+            permissions += user.group_permission_codenames
+
+        except AttributeError:
+            permissions = (
+                Permission.objects.filter(Q(user=user) | Q(group__user=user))
```

Comment:
I don't think that the Permission model has a group attribute :thinking: 
I think that you cannot get the permissions of the user that come from the group directly from the permissions, you may have to get the groups first and then the permissions associated to those groups :thinking:

Reply:
```python
In [17]: Permission.objects.filter(user=user)
Out[17]: <QuerySet [<Permission: phonedevice | Can view phone device>]>

In [18]: Permission.objects.filter(group__user=user)
Out[18]: <QuerySet [<Permission: accounts | gebruiker | Can start destruction>]>
```

![Scherm­afbeelding 2024-12-13 om 10 18 15](https://github.com/user-attachments/assets/7b416bbf-c8ed-4af4-887d-c7b4788f195d)
![Scherm­afbeelding 2024-12-13 om 10 18 06](https://github.com/user-attachments/assets/581e2bcc-f77f-47a5-9fdf-b43488b83057)